# Final Exam

For question one I have decided to use the XGBoost machine learning algorithm because it seems to be a popular algorithm for machine learning especially for classification. It produces a prediction model in the form of an ensemble of weaker prediction models , in this case decision tree models.

In [54]:
# import the libraries for panda and for xgboost and from scikit , the accuracy score function
import pandas
import xgboost
from sklearn.metrics import accuracy_score

In [55]:
# the train test split function will split the data into tran and test
from sklearn.cross_validation import train_test_split

In [57]:
#import the data as a pandas dataframe
data = pandas.read_csv('cancer_data.csv', sep=',')
dataset = data.values

In [83]:
#set the features and assign to X, while the target is Y
X = dataset[:,0:10]
Y = dataset[:,10]

In [89]:
#split the data into training, 70 percent and test 30 percent
test_size = 0.3
X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=test_size, random_state=10)

In [85]:
#fit model with no training data
model = xgboost.XGBClassifier()
model.fit(X_train, y_train)
print(model)


XGBClassifier(base_score=0.5, colsample_bylevel=1, colsample_bytree=1,
       gamma=0, learning_rate=0.1, max_delta_step=0, max_depth=3,
       min_child_weight=1, missing=None, n_estimators=100, nthread=-1,
       objective='binary:logistic', reg_alpha=0, reg_lambda=1,
       scale_pos_weight=1, seed=0, silent=True, subsample=1)


In [90]:
#make a prediction for test data
y_pred = model.predict(X_test)


In [91]:
#test accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy: %.2f%%" % (accuracy * 100.0))

Accuracy: 97.66%


In [96]:
#Test to see the results on the test data
y_pred

array([ 0.,  1.,  1.,  0.,  1.,  1.,  1.,  1.,  0.,  0.,  1.,  1.,  0.,
        1.,  0.,  0.,  1.,  1.,  1.,  1.,  0.,  0.,  1.,  1.,  0.,  0.,
        1.,  0.,  1.,  1.,  1.,  1.,  0.,  1.,  1.,  1.,  1.,  1.,  1.,
        1.,  1.,  1.,  1.,  0.,  1.,  0.,  0.,  1.,  0.,  1.,  1.,  1.,
        0.,  1.,  1.,  0.,  0.,  1.,  1.,  1.,  0.,  1.,  0.,  0.,  1.,
        1.,  1.,  1.,  0.,  0.,  1.,  1.,  1.,  1.,  1.,  0.,  1.,  1.,
        0.,  0.,  1.,  1.,  0.,  1.,  1.,  1.,  1.,  0.,  1.,  1.,  1.,
        1.,  0.,  1.,  1.,  1.,  0.,  1.,  0.,  0.,  0.,  0.,  1.,  1.,
        1.,  1.,  0.,  0.,  1.,  1.,  1.,  1.,  0.,  0.,  0.,  1.,  0.,
        0.,  0.,  1.,  1.,  0.,  1.,  0.,  1.,  1.,  0.,  0.,  1.,  1.,
        1.,  1.,  0.,  1.,  1.,  0.,  1.,  0.,  0.,  0.,  1.,  1.,  0.,
        1.,  1.,  0.,  1.,  1.,  1.,  1.,  0.,  1.,  0.,  1.,  1.,  1.,
        0.,  1.,  1.,  0.,  1.,  1.,  1.,  1.,  1.,  0.,  0.,  1.,  1.,
        1.,  1.])

In [92]:
#create a new array using the ones provided to test whether cancer is predicted ,1 or not 0
x_new=([[14,14,88,566,1,.08,.06,.04,.18,.05]])

In [93]:
#use the model given to set the prediction to y_pred2
y_pred2=model.predict(x_new)

In [94]:
#display the answer
y_pred2

array([ 1.])

As you can see the answer to qustion 1 of the exam shows a 1, which it predicts cancer based on the training data.


## Question 2
For this part i will run the code line by line then correct the bugs as i go and then answer the questions to the best of my abilities.

In [1]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
#no errors for these imports

In [3]:
from sklearn.pipeline import Pipeline
#error received, changed pipeline to Pipeline

In [5]:
from sklearn.linear_model import SGDClassifier
#error received , changed sklearn_linear_model to sklearn.linear_model

In [6]:
categories = [ 'rec.sport.baseball','rec.sport.hockey']
#no errors

In [8]:
twenty_train = fetch_20newsgroups(subset='train', categories=categories)
twenty_test = fetch_20newsgroups(subset='test', categories=categories, shuffle=True, random_state=42)
#no errors received but i can see the subsets our not labeled properly so i switched them

In [9]:
text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-3, random_state=42)),
])
#no errors, looks fine

In [10]:
text_clf.fit(twenty_train.data, twenty_train.target)
#received a mismatch error, i can see that it was fitting the test and the training together
#i switched it from twenty_test.target to twenty_train.target

Pipeline(steps=[('vect', CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip...     penalty='l2', power_t=0.5, random_state=42, shuffle=True, verbose=0,
       warm_start=False))])

In [11]:
predicted = text_clf.predict(twenty_test.data)
#no errors

In [12]:
#test to see predicted by printing
predicted

array([0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0,
       0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1,
       0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0,
       0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       1, 0,

Looks to see we have a binary classification. 1 being baseball and 0 being hockey.


1. How many observations are in the training dataset and how many features?

In [18]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()
vectors = vectorizer.fit_transform(twenty_train.data)
vectors.shape

(1197, 18571)

 i see there is 18571 columns less one being the target variable making 18570 features

In [19]:
1197*18570

22228290

In [ ]:
it seems as though there are 22228290 observations.

In [17]:
import numpy as np
np.mean((predicted == twenty_test.target)*100)
#take the average the predicted equals the target

96.859296482412063

In [20]:
#another way to check accuracy
from sklearn import metrics
metrics.f1_score(twenty_test.target, predicted, average='macro')

0.96859172557691009

looks like it performed with 97 percent accuracy .